rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/BaseLanguageServer.java

```diff
@@ -260,5 +274,120 @@ public void connect(LanguageClient client) {
+        }
+
+        private static FileChangeType convertFileChangeType(ISourceLocationChangeType changeType) throws IOException {
+            switch (changeType) {
+                case CREATED:
+                    return FileChangeType.Created;
+                case DELETED:
+                    return FileChangeType.Deleted;
+                case MODIFIED:
+                    return FileChangeType.Changed;
+                default:
+                    throw new IOException("unknown change type: " + changeType);
+            }
+        }
+
+        @Override
+        public CompletableFuture<FileStat> stat(URIParameter uri) throws IOException, URISyntaxException {
+            ISourceLocation loc = uri.getLocation();
+            return CompletableFuture.completedFuture(new FileStat(
+                reg.isDirectory(loc) ? FileType.Directory : FileType.File,
+                reg.created(loc),
+                reg.lastModified(loc),
+                reg.supportsReadableFileChannel(loc)
+                    ? reg.getReadableFileChannel(loc).size()
+                    : Prelude.__getFileSize(IRascalValueFactory.getInstance(), loc).longValue()
+                    ));
+        }
+
+        @Override
+        public CompletableFuture<FileWithType[]> readDirectory(URIParameter uri) throws URISyntaxException, IOException {
+            ISourceLocation loc = uri.getLocation();
+            return CompletableFuture.completedFuture(Arrays.stream(reg.list(loc))
+                .map(l -> new FileWithType(URIUtil.getLocationName(l), reg.isDirectory(l) ? FileType.Directory : FileType.File))
+                .toArray(FileWithType[]::new));
+        }
+
+        @Override
+        public CompletableFuture<Void> createDirectory(URIParameter uri) throws IOException, URISyntaxException {
+            ISourceLocation loc = uri.getLocation();
+            reg.mkDirectory(loc);
+            return CompletableFuture.completedFuture(null);
+        }
+
+        @Override
+        public CompletableFuture<LocationContent> readFile(URIParameter uri) throws URISyntaxException {
+            ISourceLocation loc = uri.getLocation();
+            return CompletableFuture.completedFuture(new LocationContent(Prelude.readFile(IRascalValueFactory.getInstance(), false, loc).getValue()));
+        }
+
+        @Override
+        public CompletableFuture<Void> writeFile(WriteFileParameters params) throws URISyntaxException, IOException {
+            ISourceLocation loc = params.getLocation();
+
+            if (!reg.exists(loc) && !params.isCreate()) {
+                throw new FileNotFoundException(loc.toString());
+            }
+
+            if (!reg.exists(URIUtil.getParentLocation(loc)) && params.isCreate()) {
+                throw new FileNotFoundException(URIUtil.getParentLocation(loc).toString());
+            }
+
+            if (reg.exists(loc) && params.isCreate() && !params.isOverwrite()) {
```

Comment:
file checks like "exist"  are a bit expensive, we could ask it once at the top of the method to reduce IO.